src/speculators/models/eagle.py

```diff
@@ -112,6 +169,351 @@ def validate_transformer_layer_config(cls, value: Any) -> PretrainedConfig:
+        )
+
+        # Initialize layers based on the configuration
+        self.embedding_layernorm: Optional[nn.Module] = self._create_layernorm()
+        self.fusion_fc: nn.Linear = nn.Linear(
+            2 * self.hidden_size,
+            self.hidden_size,
+            bias=config.fusion_bias,
+        )
+        self.transformer: nn.Module = self._create_transformer_layer()
+        self.pre_lm_head_layernorm: Optional[nn.Module] = self._create_layernorm()
+
+        self.post_init()  # type: ignore[attr-defined]
+
+    def attach_verifier(
+        self,
+        verifier: Union[str, os.PathLike, PreTrainedModel],
+        mode: Optional[Literal["full", "train_only"]] = None,
+    ) -> PreTrainedModel:
+        """
+        Attach a verifier model to the EagleSpeculator for speculative decoding.
+        Utilizes the verifier's embed_tokens, rotary_emb, and lm_head layers
+        for the speculator's forward pass and generation methods.
+        Additionally, for `generate`, it uses the verifier's hidden states
+        to generate speculative token predictions.
+
+        If mode is "full", the verifier is fully integrated for use with
+        both `generate` and `forward` methods.
+
+        If mode is "train_only", only the verifier's layers required for a forward pass
+        are attached, allowing for better resource utilization during training.
+        `generate` will not be available until a full verifier is attached.
+
+        Example:
+            ```python
+            # Load and attach a verifier
+            verifier = EagleSpeculator(...)
+
+            # For generation
+            speculator.attach_verifier(verifier)
+            outputs = speculator.generate(input_ids)
+            speculator.detach_verifier()
+
+            # For training
+            speculator.attach_verifier(verifier, mode="train_only")
+            outputs = speculator(input_ids, hidden_states)
+            speculator.detach_verifier()
+            ```
+
+        :param verifier: The verifier model to attach. This can be a path to a local
+            model directory, a Hugging Face model identifier, or an instance of
+            PreTrainedModel. If a path or identifier is provided, the model will be
+            loaded automatically. If an instance is provided, it will be used directly.
+        :param mode: The mode for attaching the verifier. Can be "full" or "train_only".
+            If None, defaults to "full". In "train_only" mode, only the layers
+            required for a forward pass are attached, and the speculator cannot
+            perform generation until a full verifier is attached.
+        :return: The PreTrainedModel instance for the verifier that was attached.
+        """
+        verifier = super().attach_verifier(
+            verifier=verifier,
+            mode=mode,
+        )
+
+        # Extract layers from the verifier model
+        self.embed_tokens = verifier.embed_tokens  # type: ignore[attr-defined]
```

Comment:
This will fail for some models. We can land as is, but we'll need to fix it. For instance, for Mistral this should look like

self.embed_tokens = verifier.model.language_model.embed_tokens
self.rotary_emb = verifier.model.language_model.rotary_emb
self.lm_head = verifier.lm_head

Reply:
yep, there's quite a bit we'll need to do to start generalizing and supporting othe rmodels. This extraction here is part of it, the forward may need to be edited, and as we support more transformer layer types, we'll need to expand those as well